Add oidc token support (#5670)
terraform-providers/terraform-provider-google: website/docs/d/datasource_google_service_account_id_token.html.markdown

---
subcategory: "Cloud Platform"
layout: "google"
page_title: "Google: google_service_account_id_token"
sidebar_current: "docs-google-service-account-id-token"
description: |-
  Produces OpenID Connect token for service accounts
---

# google\_service\_account\id\_token

> This data source provides a google OpenID Connect (`oidc`) `id_token`.  Tokens issued using this datasource are typically used to call external services that accept OIDC tokens for authentication (e.g. [Google Cloud Run](https://cloud.google.com/run/docs/authenticating/service-to-service)). 

Reviewer: ```suggestion
This data source provides a Google OpenID Connect (`oidc`) `id_token`.  Tokens issued from this data source are typically used to call external services that accept OIDC tokens for authentication (e.g. [Google Cloud Run](https://cloud.google.com/run/docs/authenticating/service-to-service)). 
```

For more information see
[OpenID Connect](https://openid.net/specs/openid-connect-core-1_0.html#IDToken).

## Example Usage - ServiceAccount JSON credential file.

>   Datasource will the service account provided by `GOOGLE_APPLICATION_CREDENTIALS` or `GOOGLE_CLOUD_KEYFILE_JSON` environment variable. 

Reviewer: ```suggestion
  `google_service_account_id_token` will use the configured [provider credentials](https://www.terraform.io/docs/providers/google/guides/provider_reference.html#credentials-1)
```

  
  ```hcl
  data "google_service_account_id_token" oidc {
    target_audience = "https://foo.bar/"
  }

  output "oidc_token" {
    value = data.google_service_account_id_token.oidc.id_token
  }
  ```

> ## Example Usage - `Application Default Credentials`

Reviewer: This is really the same as the previous example in the end- I made a suggestion for how to merge the two.

Author: must've missed the suggest; i guess you meant delete the dup so thats in here now

  Datasource will automatically use credentials provided by [Application Default Credentials](https://cloud.google.com/docs/authentication/production)

## Example Usage - Service Account Impersonation.

>   Datasource will use background impersonated credentials provided by [google_service_account_access_token](https://www.terraform.io/docs/providers/google/d/datasource_google_service_account_access_token.html).  

Reviewer: ```suggestion
  `google_service_account_id_token` will use background impersonated credentials provided by [google_service_account_access_token](https://www.terraform.io/docs/providers/google/d/datasource_google_service_account_access_token.html).  
```

>   Note: to use the following, you must grant `target_service_account` the

Reviewer: Does it make sense to have this as an iam_member resource in the example?

Author: i think that part is included in the docs about impersonation datasource though. ...it'll make it a bit long here to include the full flow.  IMO, the impersonation thing isn't that common vs plain svc account id token stuff

  `roles/iam.serviceAccountTokenCreator` role on itself.

  ```hcl
  data "google_service_account_access_token" "impersonated" {
    provider = google
    target_service_account = "impersonated-account@project.iam.gserviceaccount.com"
    delegates = []
    scopes = ["userinfo-email", "cloud-platform"]
    lifetime = "300s"
  }

  provider google {
    alias  = "impersonated"
    access_token = data.google_service_account_access_token.impersonated.access_token
  }

  data "google_service_account_id_token" oidc {
    provider = google.impersonated
    target_service_account = "impersonated-account@project.iam.gserviceaccount.com"
    delegates = []
    include_email = true
    target_audience = "https://foo.bar/"
  }

  output "oidc_token" {
    value = data.google_service_account_id_token.oidc.id_token
  }
  ```

## Example Usage - Invoking Cloud Run Endpoint

  The following configuration will invoke [Cloud Run](https://cloud.google.com/run/docs/authenticating/service-to-service) endpoint where the service account for Terraform has been granted `roles/run.invoker` role previously.

```hcl

> data "google_service_account_id_token" oidc {

Reviewer: ```suggestion
data "google_service_account_id_token" "oidc" {
```

  target_audience = "https://your.cloud.run.app/"
}

> data "http" cloudrun {

Reviewer: ```suggestion
data "http" "cloudrun" {
```

>   url = "https://your.cloud.run.app/"

Reviewer: Is there an existing cloud run resource that gets you this url?

Author: you mean like as part of the test?   
https://www.terraform.io/docs/providers/google/r/cloud_run_service.html  does create the cloud run resource here.  (eg, you use that to deploy some small webapp with authentication enabled (whcih is default i think), then use the service account the test runs as to setup iam binding (then wait maybe 60s  or so,, then use this provider to invoke that app

  request_headers  = {
    Authorization = "Bearer ${data.google_service_account_id_token.oidc.id_token}"
  }
}

> output "oidc_token" {

Reviewer: Since this example already shows the token being used, how valuable is it to also show it as an output?

Author: i can remove it but its shown there so a user can "inspect" the token too (eg, plug it into a jwt decoder to see its claims, etc).  removed here now though

  value = data.google_service_account_id_token.oidc.id_token
}

output "cloud_run_response" {
  value = data.http.cloudrun.body
}
```

## Argument Reference

The following arguments are supported:

* `target_audience` (Required) - The audience claim for the `id_token`.

> * `target_service_account` (Optional) - THe email of the service account being impersonated.  Used only when using impersonation mode.

Reviewer: ```suggestion
* `target_service_account` (Optional) - The email of the service account being impersonated.  Used only when using impersonation mode.
```

---
subcategory: "Cloud Platform"
layout: "google"
page_title: "Google: google_service_account_id_token"
sidebar_current: "docs-google-service-account-id-token"
description: |-
  Produces OpenID Connect token for service accounts
---

# google\_service\_account\id\_token

This data source provides a google OpenID Connect (`oidc`) `id_token` 

For more information see

> [OpenID Connect](https://openid.net/specs/openid-connect-core-1_0.html#IDToken) as well as [Authenticating using Google OpenID Connect Tokens](https://github.com/salrashid123/google_id_token)

Reviewer: Is there not an official GCP docs site that explains this?

Author: unfortunately, no, thats the only doc about google OIDC tokens...

we do have:
  https://developers.google.com/identity/protocols/oauth2/openid-connect
but thats just focued on oidc tokens for user flow (whcih isn't used here as such)

as an internal ref: see b/136495350

> ## Example Usage

Reviewer: It's probably worth thinking about how to be a bit more concise with the examples and descriptions of them. This is fairly wordy right now.

There are several ways to use this data source to get an `id_token`:

> * Using ServiceAccount JSON credential file.

Reviewer: Typically we ## each example individually rather than bullet points: see https://github.com/terraform-providers/terraform-provider-google/blob/master/website/docs/r/compute_address.html.markdown for an example docs  page with multiple examples

>   In this mode, a the datasource will the service account provided by `GOOGLE_APPLICATION_CREDENTIALS` or `GOOGLE_CLOUD_KEYFILE_JSON` environment variable. 

Reviewer: ```suggestion
  In this mode, the datasource will use the service account provided by `GOOGLE_APPLICATION_CREDENTIALS` or `GOOGLE_CLOUD_KEYFILE_JSON` environment variable. 
```
I think that's what this sentence was trying to say?

  
  Sample configuration:

>   ```yaml

Reviewer: ```suggestion
  ```hcl
```
(here and elsewhere)

>   provider google {}

Reviewer: We typically leave this out of examples unless the example has a specific reason for including it

  data "google_service_account_id_token" oidc {
    target_audience = "https://foo.bar/"
  }

  output "oidc_token" {
    value = data.google_service_account_id_token.oidc.id_token
  }
  ```

> * Using `Application Default Credentials` on ComputeEngine or Kubernetes Engine

Reviewer: ```suggestion
* Using `Application Default Credentials` on Compute Engine or Kubernetes Engine
```

>   In this mode, the datastource will use the the [instance identity document](https://cloud.google.com/compute/docs/instances/verifying-instance-identity) to provide the `id_token.  The sample usage is the same as the mode above except no environment variables are needed.

Reviewer: ```suggestion
  In this mode, the data source will use the the [instance identity document](https://cloud.google.com/compute/docs/instances/verifying-instance-identity) to provide the `id_token`.  The sample usage is the same as the mode above except no environment variables are needed.
```

* Using Service Account Impersonation.

>   In this mode, the service account that runs terraform will attempt to impersonate another service account and then acquire the second service accounts `id_token`.  This mode is similar to `access_token` that is provided by [google_service_account_access_token](https://www.terraform.io/docs/providers/google/d/datasource_google_service_account_access_token.html).  Utilizing this mechanism is requires IAM `TokenCreator` role to be granted to the the origin service account _on_ the target account.

Reviewer: ```suggestion
  In this mode, the service account that runs terraform will attempt to impersonate another service account and then acquire the second service account's `id_token`.  This mode is similar to `access_token` that is provided by [google_service_account_access_token](https://www.terraform.io/docs/providers/google/d/datasource_google_service_account_access_token.html).  Utilizing this mechanism requires the IAM `TokenCreator` role to be granted to the the origin service account _on_ the target account.
```

>   There are further two variations on how this mode is used depending on how the source token is acquired.

Reviewer: ```suggestion
  There are two further variations on how this mode is used depending on how the source token is acquired:
```

>   1. If the source token is provided directly as a variable using the requestors token, then grant the requestor the `TokenCreator` role on the target service account.  In the following case, the identity in context with `gcloud` should have the TokenCreator role on `impersonated-account@project.iam.gserviceaccount.com`

Reviewer: ```suggestion
  1. If the source token is provided directly as a variable using the requestor's token, then grant the requestor the `TokenCreator` role on the target service account.  In the following case, the identity in context with `gcloud` should have the TokenCreator role on `impersonated-account@project.iam.gserviceaccount.com`
```

>   Then run the terraform apply and specify the token directly

Reviewer: ```suggestion
  Then run `terraform apply` and specify the token directly
```

  >> Note: this variation should be extremely rare since you should be able to directly use the source system's service account JSON file directly.

  ```yaml
  provider google {}

  data "google_service_account_access_token" "default" {
  provider = google
  target_service_account = "impersonated-account@project.iam.gserviceaccount.com"
  delegates = []
  scopes = ["userinfo-email", "cloud-platform"]
  lifetime = "300s"
  }

  provider google {
    alias  = "impersonated"
    access_token = data.google_service_account_access_token.default.access_token
  }

  data "google_service_account_id_token" oidc {
    provider = google.impersonated
    
    target_service_account = "impersonated-account@project.iam.gserviceaccount.com"
    include_email = true

    target_audience = "https://foo.bar/"
  }

  output "oidc_token" {
    value = data.google_service_account_id_token.oidc.id_token
  }
  ```

## Argument Reference

The following arguments are supported:

> * `target_audience` (Required) - The audience claim value to to issued the `id_token` for.

Reviewer: I'm not sure what this sentence means

> * `access_token` (Optional) - Raw access_token to use as the source credential.  Specifying this value will use IAMCredentials API even if any other variable is set.

Reviewer: ```suggestion
* `access_token` (Optional) - Raw access_token to use as the source credential.  Specifying this value will use the IAMCredentials API even if any other variable is set.
```

If this  is the case, do you want to make it impossible for people to set this along with other fields? If so, you can set `ConflictsWith` in the schema

> * `target_service_account` (Optional) - The email of the service account bing impersonated. Used only when using impersonation mode.

Reviewer: ```suggestion
* `target_service_account` (Optional) - The email of the service account being impersonated. Used only when using impersonation mode.
```

> * `delegates` (Optional) - Delegate chain of approvals needed to perform full impersonation. Specify the fully qualified service account name.   Used only when using impersonation mode.

Reviewer: ```suggestion
* `delegates` (Optional) - Delegate chain of approvals needed to perform full impersonation. Specify the fully qualified service account name. Used only when using impersonation mode.
```
